From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sylvester Kaczmarek
 <16242628+sylvesterkaczmarek@users.noreply.github.com>
Date: Mon, 7 Sep 2026 23:21:55 +0100
Subject: [PATCH] 9600: Report oversized uncompressed SLA files

9575: Report oversized uncompressed SLA files
---
 .../SoftwareModeling/certification.manifest   |  1 +
 .../java/ghidra/pcode/utils/SlaFormat.java    |  7 ++-
 .../ghidra/pcode/utils/SlaFormatTest.java     | 52 +++++++++++++++++++
 3 files changed, 59 insertions(+), 1 deletion(-)
 create mode 100644 Ghidra/Framework/SoftwareModeling/src/test/java/ghidra/pcode/utils/SlaFormatTest.java

diff --git a/Ghidra/Framework/SoftwareModeling/certification.manifest b/Ghidra/Framework/SoftwareModeling/certification.manifest
index b12f378373..08c697864c 100644
--- a/Ghidra/Framework/SoftwareModeling/certification.manifest
+++ b/Ghidra/Framework/SoftwareModeling/certification.manifest
@@ -43,6 +43,7 @@ src/main/java/ghidra/program/model/util/package.html||GHIDRA||||END|
 src/main/resources/images/audio-volume-medium.png||Tango Icons - Public Domain|||tango icon set|END|
 src/main/resources/images/closedBookBlue.png||GHIDRA||||END|
 src/test.slow/java/log4j.xml||GHIDRA||reviewed||END|
+src/test/java/ghidra/pcode/utils/SlaFormatTest.java||GHIDRA||||END|
 src/test/resources/a_nestedif2.input||GHIDRA||reviewed||END|
 src/test/resources/a_nestedif2.output||GHIDRA||reviewed||END|
 src/test/resources/a_simpledefine.input||GHIDRA||reviewed||END|
diff --git a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/pcode/utils/SlaFormat.java b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/pcode/utils/SlaFormat.java
index 7860ad6dbc..9d10d82f7d 100644
--- a/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/pcode/utils/SlaFormat.java
+++ b/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/pcode/utils/SlaFormat.java
@@ -39,7 +39,7 @@ public class SlaFormat {
 	public static final int FORMAT_VERSION = 4;
 
 	/**
-	 * Absolute limit on the number of bytes in a .sla file
+	 * Absolute limit on the number of uncompressed bytes in a .sla file
 	 */
 	public static final int MAX_FILE_SIZE = 1 << 24;		// 16 Megabytes
 	// Attributes
@@ -269,6 +269,11 @@ public class SlaFormat {
 
 			try (InflaterInputStream inflaterStream = new InflaterInputStream(stream)) {
 				decoder.ingestStream(inflaterStream);
+				if (inflaterStream.read() != -1) {
+					throw new IOException(
+						"Uncompressed .sla file exceeds maximum supported size of " +
+							MAX_FILE_SIZE + " bytes");
+				}
 			}
 
 			decoder.endIngest();
diff --git a/Ghidra/Framework/SoftwareModeling/src/test/java/ghidra/pcode/utils/SlaFormatTest.java b/Ghidra/Framework/SoftwareModeling/src/test/java/ghidra/pcode/utils/SlaFormatTest.java
new file mode 100644
index 0000000000..15d399d778
--- /dev/null
+++ b/Ghidra/Framework/SoftwareModeling/src/test/java/ghidra/pcode/utils/SlaFormatTest.java
@@ -0,0 +1,52 @@
+/* ###
+ * IP: GHIDRA
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ghidra.pcode.utils;
+
+import static org.junit.Assert.*;
+
+import java.io.*;
+import java.util.zip.DeflaterOutputStream;
+
+import org.junit.Test;
+
+import generic.jar.ResourceFile;
+
+public class SlaFormatTest {
+
+	@Test
+	public void testBuildDecoderRejectsOversizedUncompressedSla() throws Exception {
+		File slaFile = File.createTempFile("oversized", ".sla");
+		slaFile.deleteOnExit();
+
+		try (OutputStream fileStream = new BufferedOutputStream(new FileOutputStream(slaFile))) {
+			SlaFormat.writeSlaHeader(fileStream);
+			try (DeflaterOutputStream compressedStream = new DeflaterOutputStream(fileStream)) {
+				byte[] buffer = new byte[8192];
+				int remaining = SlaFormat.MAX_FILE_SIZE + 1;
+				while (remaining > 0) {
+					int length = Math.min(remaining, buffer.length);
+					compressedStream.write(buffer, 0, length);
+					remaining -= length;
+				}
+			}
+		}
+
+		IOException exception = assertThrows(IOException.class,
+			() -> SlaFormat.buildDecoder(new ResourceFile(slaFile)));
+		assertEquals("Uncompressed .sla file exceeds maximum supported size of " +
+			SlaFormat.MAX_FILE_SIZE + " bytes", exception.getMessage());
+	}
+}
-- 
2.45.1

